### ISSUE TITLE
CS Example(s) for intermediate update, early return, clocks

### ISSUE THREAD

**Issue by chrbertsch** [modelica/fmi-standard#1057]
singled out from #1019 
It would be good to add Intermediate update with intermediate input and output variables and early return in 4.2.4. Code Example for Basic Co-Simulation

**Comment by chrbertsch** [comment]
@t-sommer : your branch https://github.com/t-sommer/Reference-FMUs/tree/cs-events-example addresses this issue, right?

**Comment by masoud-najafi** [comment]
I prepared an example but had issue with submitting on Reference-FMUs github. Then since @t-sommer had already prepared a similar example, I gave up submitting.

**Comment by t-sommer** [comment]
> @t-sommer : your branch https://github.com/t-sommer/Reference-FMUs/tree/cs-events-example addresses this issue, right?

The above example demonstrates clocks in CS and early returns, but not intermediate updates. We should create a separate example for this.

**Comment by masoud-najafi** [comment]
I noticed that there is no FMU example for synchronous clock handling. I am working on preparing a complete working example and will share it.

**Comment by chrbertsch** [comment]
Ticket-viewing-meeting:

Masoud: TorstenS is working on an example
Klaus: but not for intermediate update
Masoud: I will clarify how to commit my example with TorstenS

**Comment by masoud-najafi** [comment]
Hi @t-sommer ,
I made an example for BCS with early return without testing it. I understood that you had already done this on your side, so I abandoned my own example and did not add it to reference FMUs. Please let me know if you need me to add another example?
Thanks 
Masoud

**Comment by t-sommer** [comment]
We'll have to wait until the specification of the clocks in CS is finalized.

**Comment by chrbertsch** [comment]
It would be very important to have CS examples with the new clocks features to understand and check the improved clocks description
@t-sommer : what about your https://github.com/t-sommer/Reference-FMUs/tree/cs-events-example example? Are big changes neded before merging?
@masoud-najafi : and what about your examples?

**Comment by chrbertsch** [comment]
Regular design meeting: 

Masoud, Claudio: let's do this after the plugfest

**Comment by chrbertsch** [comment]
Web-Meeting:
TorstenB: there is an example in the sandbox (bouncing ball with events, created from SimulationX)
Pierre: would be good to have this in the reference FMUs. This is an individual FMU. Perhaps we would need connected FMUs
TorstenB: it would help to have some example code for intermediate update, accessing inputs and outputs.
Klaus: intermediate outputs with bouncing ball would be easy. We will work on implementing this.

**Comment by masoud-najafi** [comment]
Hi Torsten (@t-sommer),
CC: @chrbertsch 
I have modified the bouncingBall example of the reference FMU to support earlyReturn and EventMode for CS.  It works fine on my simulator by setting EventModeUsed to both False and True. Would you please check this changes in the reference FMU? Here are the modified files:
[BouncingBall_with_EarlyReturn.zip](https://github.com/modelica/fmi-standard/files/7549593/BouncingBall_with_EarlyReturn.zip)
Once, approved by you, we can ask @klausschuch  to check it on AVL tool.

**Comment by chrbertsch** [comment]
@t-sommer : so the early return part of this issue is done with https://github.com/modelica/Reference-FMUs/commit/f2b181e641c6e1e9ed119c5d4fc19de2f2c2031f , right?

@robbr48 : it would be great if you could provide an example for intermediate update!

**Comment by andreas-junghanns** [comment]
> @robbr48 : it would be great if you could provide an example for intermediate update!

In Reference-FMUs/examples/cs_intermediate_update.c you have one now.

Can we close this issue now with all the place examples have been provided?

**Comment by chrbertsch** [comment]
>> @robbr48 : it would be great if you could provide an example for intermediate update!

>In Reference-FMUs/examples/cs_intermediate_update.c you have one now.

This examples shows only intermediate outputs; it would be great to have an example also covering intermediate inputs to closer to advanced co-simulation applications. @robbr48 : do you think you can provide one?

> Can we close this issue now with all the place examples have been provided?

We are still missing an example for CS with Clocks. Do we see a chance to get one?

**Comment by masoud-najafi** [comment]
Hi @t-sommer,

I checked the code used in the reference FMU. Maybe I'm missing something, but
1) if eventModeUsed      is set to False, DoStep should never return with eventEncountered=True
2) if earlyReturnAllowed is set to False, DoStep should never return with earlyReturn=True.
This is not guaranteed in the fmi3DoStep code.

Also, if in the XML, the FMU indicates canReturnEarlyAfterIntermediateUpdate="true" and  hasEventMode="true", 
the FMU should be able to work with four different instantiation settings, i.e.,
eventModeUsed=FALSE , earlyReturnAllowed=FALSE  
eventModeUsed=TRUE , earlyReturnAllowed=FALSE    
eventModeUsed=FALSE , earlyReturnAllowed=TRUE
eventModeUsed=TRUE , earlyReturnAllowed=TRUE

I checked the bouncing-ball example in reference FMUs, and if I am not wrong, it looks like only the last case works correctly.
